src/main/scala/linting/rule/LintConflictingModuleNames.scala

```diff
@@ -0,0 +1,314 @@
+      case ext: ExtModule => ext
+    }
+    val main = nameMappings.getOrElse(circuit.main, circuit.main)
+    circuit.copy(main = main, modules = modules)
+  }
+}
+
+/** A strategy for generating a new names for modules
+  */
+sealed trait NamingStrategy {
+
+  /** Generates a new stable module name based on the desired name and the module IR node
+    *
+    * called by [[LintConflictingModuleNames]] to rename modules
+    *
+    * @param desiredName the requested name for the module, generated name should contain this name
+    * @param module the module targeted by the desiredName
+    */
+  def getName(desiredName: String)(module: Module): String
+}
+
+/** A naming strategy that always returns the desired name
+  */
+case object ExactNamingStrategy extends NamingStrategy {
+  def getName(desiredName: String)(module: Module): String = desiredName
+}
+
+/** Specifies a naming strategy to use for a module and the modules that it collides with
+  */
+case class NamingStrategyAnnotation(
+  strategy: NamingStrategy,
+  target: IsModule
+) extends SingleTargetAnnotation[IsModule] with HasSerializationHints {
+  def typeHints: Seq[Class[_]] = Seq(
+    ExactNamingStrategy.getClass
+  )
+
+  def duplicate(newTarget: IsModule): NamingStrategyAnnotation = {
+    this.copy(target = newTarget)
+  }
+}
+
+/** Specifies the original desired name for a module
+  */
+case class DesiredNameAnnotation(
+  desiredName: String,
+  target: IsModule
+) extends SingleTargetAnnotation[IsModule] {
+  def duplicate(newTarget: IsModule): DesiredNameAnnotation = {
+    this.copy(target = newTarget)
+  }
+}
+
+/** Specifies the desired name to rename the module to
+  */
+case class OverrideDesiredNameAnnotation(
+  desiredName: String,
+  target: IsModule
+) extends SingleTargetAnnotation[IsModule] {
+  def duplicate(newTarget: IsModule): OverrideDesiredNameAnnotation = {
+    this.copy(target = newTarget)
+  }
+}
+
+/** Collects [[DesiredNameAnnotation]]s for [[LintConflictingModuleNames]] to lint
+  */
+case object LintConflictingModuleNamesAspect extends Aspect[RawModule] {
+  def toAnnotation(top: RawModule): AnnotationSeq = {
+    Select.collectDeep(top) {
+      case m: chisel3.Module => DesiredNameAnnotation(m.desiredName, m.toTarget)
+    }.toSeq
+  }
+}
+
+/** An example of how to customize the names of certain modules using an Aspect
+  */
+case object StabilizeNamesAspect extends Aspect[RawModule] {
+  def toAnnotation(top: RawModule): AnnotationSeq = {
+    RunFirrtlTransformAnnotation(new RenameDesiredNames) +:
+    Select.collectDeep(top) {
+      // annotating all Queues with a more descriptive desired name
+      case m: Queue[_] => Seq(
+        new OverrideDesiredNameAnnotation(s"Queue_${m.genType.getClass.getSimpleName}_entries_${m.entries}", m.toTarget)
+      )
+      case m: TLMonitor => Seq(
+        new OverrideDesiredNameAnnotation(m.desiredName, m.toTarget)
```

Comment:
this is basically a NOP, right?

Reply:
this will still try to rename all `TLMonitor_*` to `TLMonitor` and will most likely fail because they don't dedup.

Reply:
but wouldn't they fail because they don't dedup anyway without this annotation? Isn't their desired name already going to be TLMonitor?

Reply:
without this annotation the names will stay as `TLMonitor_*` and will not fail. Renaming is only attempted if there is an `OverrideDesiredNameAnnotation`. any names not targeted by an annotation are not affected. the chisel `desiredName` is TLMonitor but chisel will emit modules named TLMonitor_1 TLMonitor_2 ... and the chisel desired name does not get passed to firrtl unless there is an annotation

Reply:
I thought that the linting pass would apply `DesiredNameAnnotation` to all modules so that the linting could be performed...?

Reply:
the aspect provides the `DesiredNameAnnotations`. the firrtl linting pass just checks them

Reply:
Wouldn't a better (more realistic) example here just be `DesiredNameAnnotation`? By the time the linter is running, isn't it only looking at the DesiredNameAnnotations (because the OverrideDesiredName annotations have been turned into DesiredNameAnnotations by an earlier transform)?